### PR TITLE
Chisel Pipelined GEMM

### PR DESCRIPTION
**Pull request by adavare**
A CHISEL pipelined GEMM module along with associated unit tests are being added with this pull request. 

Significant cycle count improvements are seen for conv2D workloads. In particular, when evaluated with "tvm/vta/tutorials/frontend/deploy_classification.py" with the tsim target using all 6 supported ResNets, overall speedups range from 3.10x to 3.69x as shown in the table below. Top 5 labels all match for the default "cat.png" image in all cases.

| Net          | Baseline, cycles | With Pipelined GEMM, cycles  | Speedup |
| :------------: | --------------------: | -----------------: | -------: |
| resnet18_v1  | 27,679,515           | 7,729,601         |    3.58 |
| resnet34_v1  | 57,471,735           | 15,736,441        |    3.65 |
| resnet18_v2  | 27,558,980           | 7,608,808         |    3.62 |
| resnet34_v2  | 57,270,321           | 15,534,528        |    3.69 | 
| resnet50_v2  | 68,187,407           | 22,012,906        |    3.10 |
| resnet101_v2 | 129,263,035          | 39,182,158        |    3.30 |

Code contributions to this PR were made by the following individuals (in alphabetical order): @suvadeep89, @stevenmburns, @pasqoc, @adavare, @sjain12intel, @aasorokiin, and @zhenkuny.

### CODE REVIEW

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 1]
```diff
@@ -421,3 +544,218 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
+    // if (io.dec.empty_1 != None) assert(io.dec.empty_1.get === 0.U)
+  }.elsewhen( state === sRun && m.io.last) {
+    state := sWait
+  }.elsewhen( state === sWait && inflight === 0.U) {
+    state := sIdle
+    io.done := true.B
+  }
+  io.state := state
+
+  if ( false) {
```
Comment (stevenmburns) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: @adavare We don't need this printf debug anymore.

Comment (adavare) [reply hardware/chisel/src/main/scala/core/TensorGemm.scala]: @stevenmburns: Removed, thanks for catching this!

File: hardware/chisel/src/test/scala/unittest/TensorGemmJsonTest.scala [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+  println(s"Total active steps: ${total_steps}")
+  mocks.test_if_done()
+}
+
+class TensorGemmJsonTestx extends GenericTest("TensorGemmJson", (p:Parameters) => new TensorGemmPipelinedSplit()(p),
+  (c:TensorGemmPipelinedSplit) => new TensorGemmJsonTester(c, "/x.json"))
+
+class TensorGemmJsonTesty extends GenericTest("TensorGemmJson", (p:Parameters) => new TensorGemmPipelinedSplit()(p),
+  (c:TensorGemmPipelinedSplit) => new TensorGemmJsonTester(c, "/y.json"))
```
Comment (vegaluisjose) [inline hardware/chisel/src/test/scala/unittest/TensorGemmJsonTest.scala]: Hey @adavare , nice PR.

can we change `x.json` and `y.json` to have more meaningful names? It looks like these are test-data...is there some information that `x` tests versus `y` that we could use for naming?

**Review by vegaluisjose — CHANGES_REQUESTED** [review]

File: hardware/chisel/src/test/scala/unittest/TensorGemmJsonTest.scala (hunk 2 above)
Comment (adavare) [reply hardware/chisel/src/test/scala/unittest/TensorGemmJsonTest.scala]: Thanks for the suggestion, I've renamed both files.

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 3]
```diff
@@ -272,10 +395,10 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
   }.elsewhen(!dec.reset) {
     when((state === sReadTensor) && mvc.io.acc_o.data.valid) { // issue & commit
     }.elsewhen(state === sReadTensor) { // issue a tensor
-      assert(inflight =/= ((1<<inflightBits)-1).U)
+      assert( inflight =/= ((1<<inflightBits)-1).U)
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 4]
```diff
@@ -272,10 +395,10 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
       inflight := inflight + 1.U
     }.elsewhen(mvc.io.acc_o.data.valid) { // commit a tensor
-      assert(inflight =/= 0.U)
+      assert( inflight =/= 0.U)
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 5]
```diff
@@ -368,8 +491,8 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
 
   io.done := done
 
-  if (debug) {
-    printf("[TensorGemm] [state]:%d [inflight]:%d\n", state, inflight)
+  if ( debug) {
+    printf( "[TensorGemm] [state]:%d [inflight]:%d\n", state, inflight)
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 6]
```diff
@@ -421,3 +544,206 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
+  }.elsewhen( state === sRun && m.io.last) {
+    state := sWait
+  }.elsewhen( state === sWait && inflight === 0.U) {
+    state := sIdle
+    io.done := true.B
+  }
+  io.state := state
+
+  assert( state =/= sRun  || capture_dec.asUInt === io.dec.asUInt)
+  assert( state =/= sWait || capture_dec.asUInt === io.dec.asUInt)
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 7]
```diff
@@ -421,3 +544,206 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
+
+  val capture_dec = Reg( chiselTypeOf( io.dec))
+
+  io.done := false.B
+  when( state === sIdle && io.start) {
+    state := sRun
+    capture_dec := io.dec
+    // if (io.dec.empty_0 != None) assert(io.dec.empty_0.get === 0.U)
+    // if (io.dec.empty_1 != None) assert(io.dec.empty_1.get === 0.U)
+  }.elsewhen( state === sRun && m.io.last) {
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 8]
```diff
@@ -421,3 +544,206 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
+    state := sWait
+  }.elsewhen( state === sWait && inflight === 0.U) {
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 9]
```diff
@@ -421,3 +544,206 @@ class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorG
...
+    io.wgt.rd(idx).idx.bits := ShiftRegister(uop_wgt, scratchpadReadLatency)
+
+    assert(io.wgt.rd(idx).data.valid === ShiftRegister(delayed_uop_valid, scratchpadReadLatency))
+  }
+  io.wgt.tieoffWrite()
+  io.inp.tieoffWrite()
+
+  // create a pipe of 3+ delay with split by goup last stage
+  // and a separate last stage for out and inflight
+  val wrpipe0 = Module( new Pipe(chiselTypeOf(io.acc.wr(0).bits.idx), latency= 2 + scratchpadReadLatency))
```
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: extra space after open paren

**Review by vegaluisjose — CHANGES_REQUESTED** [review]

File: hardware/chisel/src/main/scala/core/TensorGemm.scala (hunk 3 above)
Comment (adavare) [reply hardware/chisel/src/main/scala/core/TensorGemm.scala]: Should all be replaced.

**Review by vegaluisjose — APPROVED** [review]
LGTM

File: hardware/chisel/src/test/scala/unittest/TensorGemmTest.scala [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+  expect(c.io.out.wr(0).valid, 0)
+  expect(c.io.acc.wr(0).valid, 0)
+
+  mocks.logical_step(0, 0)
+  expect(c.io.state, c.sExe)
+  expect(c.io.out.wr(0).valid, 0)
+  expect(c.io.acc.wr(0).valid, 0)
+  expect(c.io.done, 0)
+
+  mocks.logical_step(0, 0 /*1*/)
```
Comment (liangfu) [inline hardware/chisel/src/test/scala/unittest/TensorGemmTest.scala]: remove unused comment

File: hardware/chisel/src/main/scala/core/TensorGemm.scala [hunk 11]
```diff
@@ -199,7 +322,7 @@ abstract class TensorGemmIfc(implicit p: Parameters) extends Module {
  * Also, the TensorGemm uses the reset field in the Gemm instruction to
  * clear or zero-out the acc-scratchpad locations based on the micro-ops.
  */
-class TensorGemm(debug: Boolean = false)(implicit p: Parameters) extends TensorGemmIfc {
+class TensorGemmOrig(debug: Boolean = false)(implicit p: Parameters) extends TensorGemmIfc {
```
Comment (liangfu) [inline hardware/chisel/src/main/scala/core/TensorGemm.scala]: I think the `Orig` suffix could be better replace with something more meaningful, such as `Simple` or `Small`, indicating the the previous simple design consumes less logic elements, and it's easy to understand.

**Review by liangfu — CHANGES_REQUESTED** [review]
Thanks for the nice PR. I just left a few comments.

**Comment by liangfu** [comment]
I'm also curious about logic resource consumption for the simple GEMM and pipelined GEMM. Do you have numbers on that ?

File: hardware/chisel/src/test/scala/unittest/TensorGemmTest.scala (hunk 10 above)
Comment (adavare) [reply hardware/chisel/src/test/scala/unittest/TensorGemmTest.scala]: Removed

File: hardware/chisel/src/main/scala/core/TensorGemm.scala (hunk 11 above)
Comment (adavare) [reply hardware/chisel/src/main/scala/core/TensorGemm.scala]: Replaced TensorGemmOrig with TensorGemmSimple

**Comment by aasorokiin** [comment]
As for logic resource. After synthesis the pipelined VTAShell design has 43 more sequential cells. Pipelined ALU has 6% larger area. Pipelined GEMM has the same area as the Simple one.  batch=1, blockin=16, blockout=16

**Review by liangfu — APPROVED** [review]
Thanks for clarification on logic resource. The changes LGTM.

**Pull request merged**